# Notes to Improve Workflow

In [130]:
# Setup matplotlib to plot inline (within the notebook)
%matplotlib inline
# Import the pyplot module of Matplotlib as plt
import matplotlib.pyplot as plt
# Import pandas under the abbreviation 'pd'
import pandas as pd
# Import NumPy under the abbreviation 'np'
import numpy as np

#Notes for efficienccy:
#shift + tab when cursor is in arguments to bring up its documentation
#ex: pd.DataFrame(Shift+tab here)
#recall: Json == Dictionary
#to know functions of pandas:
dir(pd)

['BooleanDtype',
 'Categorical',
 'CategoricalDtype',
 'CategoricalIndex',
 'DataFrame',
 'DateOffset',
 'DatetimeIndex',
 'DatetimeTZDtype',
 'ExcelFile',
 'ExcelWriter',
 'Flags',
 'Float32Dtype',
 'Float64Dtype',
 'Float64Index',
 'Grouper',
 'HDFStore',
 'Index',
 'IndexSlice',
 'Int16Dtype',
 'Int32Dtype',
 'Int64Dtype',
 'Int64Index',
 'Int8Dtype',
 'Interval',
 'IntervalDtype',
 'IntervalIndex',
 'MultiIndex',
 'NA',
 'NaT',
 'NamedAgg',
 'Period',
 'PeriodDtype',
 'PeriodIndex',
 'RangeIndex',
 'Series',
 'SparseDtype',
 'StringDtype',
 'Timedelta',
 'TimedeltaIndex',
 'Timestamp',
 'UInt16Dtype',
 'UInt32Dtype',
 'UInt64Dtype',
 'UInt64Index',
 'UInt8Dtype',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__docformat__',
 '__file__',
 '__getattr__',
 '__git_version__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_config',
 '_hashtable',
 '_is_numpy_dev',
 '_lib',
 '_libs',
 '_np_version_under1p18',
 '_testing',
 '_tslib',
 '_typing',
 

# Reading The Data

In [131]:
# using pandas and read_csv()
acquis = pd.read_csv("datasets/startups/acquisitions.csv")
additions = pd.read_csv("datasets/startups/additions.csv")
companies = pd.read_csv("datasets/startups/companies.csv")
invests = pd.read_csv("datasets/startups/investments.csv")
rounds = pd.read_csv("datasets/startups/rounds.csv")
dfAll = [acquis, additions, companies, invests, rounds]

In [132]:
#shapes (rows x cols) of each dataset
for ele in dfAll:
    print(ele.shape)

(18969, 18)
(2213, 4)
(66368, 14)
(168647, 18)
(114949, 12)


# Samples of the datasets

In [133]:
acquis.sample(5).transpose()

5294   \
company_permalink          /organization/ecloud-collect-data-collect-pro-   
company_name                            eCloud Collect (Data Collect Pro)   
company_category_list                                                 NaN   
company_country_code                                                  NaN   
company_state_code                                                    NaN   
company_region                                                        NaN   
company_city                                                          NaN   
acquirer_permalink                                /organization/zapproved   
acquirer_name                                                   Zapproved   
acquirer_category_list  Cloud Computing|Collaboration|Ediscovery|Produ...   
acquirer_country_code                                                 USA   
acquirer_state_code                                                    OR   
acquirer_region                                          Portland, Oregon   
acquirer_city                                                    Portland   
acquired_at                                                      1/1/2014   
acquired_month                                                    2014-01   
price_amount                                                          NaN   
price_currency_code                                                   USD   

                                              2985   \
company_permalink       /organization/cardboardfish   
company_name                          CardBoardFish   
company_category_list         Service Providers|SMS   
company_country_code                            GBR   
company_state_code                               G5   
company_region                                 Kent   
company_city                                   Kent   
acquirer_permalink              /organization/mblox   
acquirer_name                                 Mblox   
acquirer_category_list          Mobile|SMS|Wireless   
acquirer_country_code                           USA   
acquirer_state_code                              CA   
acquirer_region                         SF Bay Area   
acquirer_city                              Campbell   
acquired_at                               7/14/2014   
acquired_month                              2014-07   
price_amount                                    NaN   
price_currency_code                             USD   

                                                    6433   \
company_permalink       /organization/forest-laboratories   
company_name                          Forest Laboratories   
company_category_list           Biotechnology|Health Care   
company_country_code                                  USA   
company_state_code                                     NY   
company_region                              New York City   
company_city                                     New York   
acquirer_permalink                  /organization/actavis   
acquirer_name                                     Actavis   
acquirer_category_list      Biotechnology|Pharmaceuticals   
acquirer_country_code                                 USA   
acquirer_state_code                                    NJ   
acquirer_region                                    Newark   
acquirer_city                                  Parsippany   
acquired_at                                     2/18/2014   
acquired_month                                    2014-02   
price_amount                                25000000000.0   
price_currency_code                                   USD   

                                                      4970   \
company_permalink       /organization/doctor-driven-systems   
company_name                          Doctor Driven Systems   
company_category_list                                   NaN   
company_country_code                                    USA   
company_state_code                                       MA   
company_region      

In [134]:
additions.sample(5).transpose()

699           1994                    1776                1068  \
content       IPOs  Rounds: seed  Rounds: private_equity  Profiles: Products   
month_str  2012-12       2014-05                 2010-03             2009-01   
year_str      2012          2014                    2010                2009   
value           11          1326                      13                 720   

                               1908  
content    Rounds: secondary_market  
month_str                   2015-10  
year_str                       2015  
value                             5

In [135]:
companies.sample(5).transpose()

15180                             46531  \
permalink            /organization/delphx  /organization/quintel-technology   
name                               DelphX                Quintel Technology   
homepage_url       http://www.delphx.com/       http://quintelsolutions.com   
category_list                     Finance                            Mobile   
funding_total_usd                       -                          10000000   
status                          operating                         operating   
country_code                          USA                               USA   
state_code                             PA                                NY   
region                       Philadelphia               Rochester, New York   
city                              Malvern                         Rochester   
funding_rounds                          1                                 1   
founded_at                      3/22/2011                          1/1/2007   
first_funding_at                6/17/2015                         4/14/2014   
last_funding_at                 6/17/2015                         4/14/2014   

                                              49316  \
permalink                     /organization/saguaro   
name                                        Saguaro   
homepage_url                                    NaN   
category_list      Collaboration|Internet|Messaging   
funding_total_usd                                 -   
status                                    operating   
country_code                                    NaN   
state_code                                      NaN   
region                                          NaN   
city                                            NaN   
funding_rounds                                    2   
founded_at                                      NaN   
first_funding_at                           1/1/2015   
last_funding_at                           1/10/2015   

                                                            56680  \
permalink                                  /organization/teamnote   
name                                                     TeamNote   
homepage_url                          http://www.teamnoteapp.com/   
category_list      Collaboration|Enterprise Application|Messaging   
funding_total_usd                                          161250   
status                                                  operating   
country_code                                                  NaN   
state_code                                                    NaN   
region                                                        NaN   
city                                                          NaN   
funding_rounds                                                  2   
founded_at                                               1/1/2014   
first_funding_at                                        12/1/2014   
last_funding_at                                          8/1/2015   

                                         36106  
permalink              /organization/minutrade  
name                                 MinuTrade  
homepage_url       http://www.minutrade.com.br  
category_list                              NaN  
funding_total_usd                            -  
status                               operating  
country_code                               NaN  
state_code                                 NaN  
region                                     NaN  
city                                       NaN  
funding_rounds                               1  
founded_at                                 NaN  
first_funding_at                      4/1/2013  
last_funding_at                       4/1/2013

In [136]:
invests.sample(5).transpose()

83967   \
company_permalink                            /organization/loxo-oncology   
company_name                                               Loxo Oncology   
company_category_list                                      Biotechnology   
company_country_code                                                 USA   
company_state_code                                                    NY   
company_region                                             New York City   
company_city                                                    New York   
investor_permalink                         /organization/aisling-capital   
investor_name                                            Aisling Capital   
investor_country_code                                                USA   
investor_state_code                                                   NY   
investor_region                                            New York City   
investor_city                                                   New York   
funding_round_permalink  /funding-round/95d3381fa0830540fd5ce6d6e9438001   
funding_round_type                                               venture   
funding_round_code                                                     B   
funded_at                                                       5/6/2014   
raised_amount_usd                                             24000000.0   

                                                                  38678   \
company_permalink                                  /organization/daalder   
company_name                                                     Daalder   
company_category_list               Financial Services|Payments|Software   
company_country_code                                                 NLD   
company_state_code                                                     7   
company_region                                                 Amsterdam   
company_city                                                   Amsterdam   
investor_permalink                                  /organization/keadyn   
investor_name                                                     Keadyn   
investor_country_code                                                NLD   
investor_state_code                                                   11   
investor_region                                                Rotterdam   
investor_city                                                  Rotterdam   
funding_round_permalink  /funding-round/373ceeae2a2c73998af428f5a079fcbd   
funding_round_type                                                  seed   
funding_round_code                                                   NaN   
funded_at                                                      9/22/2015   
raised_amount_usd                                               448063.0   

                                                                  63440   \
company_permalink                            /organization/harvest-power   
company_name                                               Harvest Power   
company_category_list                                   Clean Technology   
company_country_code                                                 USA   
company_state_code                                                    MA   
company_region                                                    Boston   
company_city                                                     Waltham   
investor_permalink          /organization/kleiner-perkins-caufield-byers   
investor_name                           Kleiner Perkins Caufield & Byers   
investor_country_code                                                USA   
investor_state_code                                                   CA   
investor_region                                              SF Bay Area   
investor_city                                                 Menlo Park   
funding_round_permalink  /funding-round/c8b54be8c446c96737bd1980b73c6a26   
funding_round_type                  

In [137]:
rounds.sample(5).transpose()

20248  \
company_permalink                        /organization/classic-drive-ins   
company_name                                               Classic Drive   
company_category_list                    Communities|Startup Histrionics   
company_country_code                                                 CHE   
company_state_code                                                    25   
company_region                                                    Zurich   
company_city                                                      Zürich   
funding_round_permalink  /funding-round/68083c1f91a0296c89c710d1b59cf8eb   
funding_round_type                                   equity_crowdfunding   
funding_round_code                                                   NaN   
funded_at                                                     11/12/2010   
raised_amount_usd                                                    NaN   

                                                                   81696  \
company_permalink             /organization/real-girls-media-network-inc   
company_name                                    Real Girls Media Network   
company_category_list                                           Software   
company_country_code                                                 USA   
company_state_code                                                    CA   
company_region                                               SF Bay Area   
company_city                                               San Francisco   
funding_round_permalink  /funding-round/cfa38b39528edf59c3f799235df3e170   
funding_round_type                                               venture   
funding_round_code                                                   NaN   
funded_at                                                     11/15/2008   
raised_amount_usd                                              7196231.0   

                                                                     6328   \
company_permalink                               /organization/apogeeinvent   
company_name                                                  ApogeeInvent   
company_category_list    Design|E-Commerce|Internet|SEO|Software|Web De...   
company_country_code                                                   USA   
company_state_code                                                      ID   
company_region                                                 Idaho Falls   
company_city                                                American Falls   
funding_round_permalink    /funding-round/57ec3991411a22d9c79e435d1fa7078f   
funding_round_type                                                   angel   
funding_round_code                                                     NaN   
funded_at                                                         1/8/2013   
raised_amount_usd                                                  25000.0   

                                                                     80775  \
company_permalink                                      /organization/quova   
company_name                                                         Quova   
company_category_list    Advertising|Analytics|Digital Rights Managemen...   
company_country_code                                                   NLD   
company_state_code                                                       7   
company_region                                                   Amsterdam   
company_city                                                     Amsterdam   
funding_round_permalink    /funding-round/3f37ac4cba925e94d66400bebf3c32c5   
funding_round_type                                                 venture   
funding_round_code                                                     NaN   
funded_at                                                        3/13/2001   
raised_amount_usd                                               21000000.0   

                                                           

# Seeing the columns in the datasets

In [138]:
def datasetsCols(dfList, datasetNames, pad=""):
    """
    Takes dataframes and returns a dataframe (df) with each column having a df's
    column names. The "pad" is to make sure all lists of columns have 
    the same length padded out with "pad"
    Example
    pad = "XX", datasetNames = ["cars", "planes"]
    dfList = [carsDF, planesDF]
    returned data frame:
            cars            planes
    0       numOfWheels     numOfWings
    1       manufacturer    manufacturer
    2       make            XX
    """
    cols = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        dictCsvs.update({datasetNames[i] : cols[i]})

    return pd.DataFrame(dictCsvs)

In [139]:
lstCsvNames = ["acquisitions", "additions", "companies", "investments", "rounds"]

dfAllCols = datasetsCols(dfAll, lstCsvNames)
dfAllCols

acquisitions  additions          companies  \
0        company_permalink    content          permalink   
1             company_name  month_str               name   
2    company_category_list   year_str       homepage_url   
3     company_country_code      value      category_list   
4       company_state_code             funding_total_usd   
5           company_region                        status   
6             company_city                  country_code   
7       acquirer_permalink                    state_code   
8            acquirer_name                        region   
9   acquirer_category_list                          city   
10   acquirer_country_code                funding_rounds   
11     acquirer_state_code                    founded_at   
12         acquirer_region              first_funding_at   
13           acquirer_city               last_funding_at   
14             acquired_at                                 
15          acquired_month                                 
16            price_amount                                 
17     price_currency_code                                 

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

# Data Cleaning

First, lets change the columns in the `companies` dataset <br>
to be the similar to all other datasets: <br>
(we're doing this in case we merge the datasets together, we want them to have the same column names)

In [140]:
companies.rename(columns = {'permalink' : 'company_permalink', 
                            'name' : 'company_name',
                            'category_list' : 'company_category_list',
                            'country_code' : 'company_country_code',
                            'state_code' : 'company_state_code',
                            'region' : 'company_region',
                            'city' : 'company_city'}, inplace = True)

In [141]:
dfAll[2] = companies
dfAllCols = datasetsCols(dfAll, lstCsvNames)
dfAllCols

acquisitions  additions              companies  \
0        company_permalink    content      company_permalink   
1             company_name  month_str           company_name   
2    company_category_list   year_str           homepage_url   
3     company_country_code      value  company_category_list   
4       company_state_code                 funding_total_usd   
5           company_region                            status   
6             company_city              company_country_code   
7       acquirer_permalink                company_state_code   
8            acquirer_name                    company_region   
9   acquirer_category_list                      company_city   
10   acquirer_country_code                    funding_rounds   
11     acquirer_state_code                        founded_at   
12         acquirer_region                  first_funding_at   
13           acquirer_city                   last_funding_at   
14             acquired_at                                     
15          acquired_month                                     
16            price_amount                                     
17     price_currency_code                                     

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

Let's start by analyzing `additions` dataset, as it has unusual format of columns

## Removing "additions" Dataset

In [142]:
additions.sample(5)

content month_str  year_str  value
93               Acquisitions   1998-12      1998      6
1512   Rounds: debt_financing   2015-11      2015    100
1746  Rounds: post_ipo_equity   2015-05      2015     20
587                      IPOs   2003-08      2003     12
745                  Profiles   2008-02      2008   2966

In [143]:
additions["content"].unique()

array(['Acquisitions', 'IPOs', 'Profiles', 'Profiles: Organizations',
       'Profiles: People', 'Profiles: Products', 'Rounds',
       'Rounds: angel', 'Rounds: convertible_note',
       'Rounds: debt_financing', 'Rounds: equity_crowdfunding',
       'Rounds: grant', 'Rounds: non_equity_assistance',
       'Rounds: post_ipo_debt', 'Rounds: post_ipo_equity',
       'Rounds: private_equity', 'Rounds: product_crowdfunding',
       'Rounds: secondary_market', 'Rounds: seed', 'Rounds: undisclosed',
       'Rounds: venture'], dtype=object)

In [144]:
dfAllCols

acquisitions  additions              companies  \
0        company_permalink    content      company_permalink   
1             company_name  month_str           company_name   
2    company_category_list   year_str           homepage_url   
3     company_country_code      value  company_category_list   
4       company_state_code                 funding_total_usd   
5           company_region                            status   
6             company_city              company_country_code   
7       acquirer_permalink                company_state_code   
8            acquirer_name                    company_region   
9   acquirer_category_list                      company_city   
10   acquirer_country_code                    funding_rounds   
11     acquirer_state_code                        founded_at   
12         acquirer_region                  first_funding_at   
13           acquirer_city                   last_funding_at   
14             acquired_at                                     
15          acquired_month                                     
16            price_amount                                     
17     price_currency_code                                     

                investments                   rounds  
0         company_permalink        company_permalink  
1              company_name             company_name  
2     company_category_list    company_category_list  
3      company_country_code     company_country_code  
4        company_state_code       company_state_code  
5            company_region           company_region  
6              company_city             company_city  
7        investor_permalink  funding_round_permalink  
8             investor_name       funding_round_type  
9     investor_country_code       funding_round_code  
10      investor_state_code                funded_at  
11          investor_region        raised_amount_usd  
12            investor_city                           
13  funding_round_permalink                           
14       funding_round_type                           
15       funding_round_code                           
16                funded_at                           
17        raised_amount_usd

In [145]:
additions.shape

(2213, 4)

In [146]:
companies.shape, rounds.shape

((66368, 14), (114949, 12))

So apparently there are no columns that could be used to join with the other datasets. <br>
Furthermore, the "value" column is too vague to be useful. <br>
Therefore, the "additions" dataset will be discarded.

In [147]:
dfAll.pop(1)
lstCsvNames.pop(1)
len(dfAll), lstCsvNames

(4, ['acquisitions', 'companies', 'investments', 'rounds'])

# Data Cleaning (Cont.)

## Checking All Missing Values

In [148]:
def datasetsNulls(dfList, datasetNames, nullCol="nulls_", pad=""):
    """
    Use this when you want to display a column of column names,
    then a column of the null values, and repeat that for each dataset.
    Returns a dataframe
    """
    cols = []
    nulls = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        nulls.append(df.isnull().sum().tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        nulls[i] += [None] *  (maxArrayLen - len(nulls[i]))
        dictCsvs[datasetNames[i]] = cols[i]
        dictCsvs[nullCol + str(i+1)] = nulls[i]
    
    return pd.DataFrame(dictCsvs)

In [149]:
dfAllNulls = datasetsNulls(dfAll, lstCsvNames)
dfAllNulls

acquisitions  nulls_1              companies  nulls_2  \
0        company_permalink        0      company_permalink      0.0   
1             company_name        1           company_name      1.0   
2    company_category_list     4984           homepage_url   5058.0   
3     company_country_code     4669  company_category_list   3148.0   
4       company_state_code     4949      funding_total_usd      0.0   
5           company_region     4911                 status      0.0   
6             company_city     4911   company_country_code   6958.0   
7       acquirer_permalink        1     company_state_code   8547.0   
8            acquirer_name        1         company_region   8030.0   
9   acquirer_category_list     2389           company_city   8028.0   
10   acquirer_country_code     1171         funding_rounds      0.0   
11     acquirer_state_code     1436             founded_at  15221.0   
12         acquirer_region     1360       first_funding_at     24.0   
13           acquirer_city     1360        last_funding_at      0.0   
14             acquired_at       13                             NaN   
15          acquired_month       15                             NaN   
16            price_amount    13957                             NaN   
17     price_currency_code        7                             NaN   

                investments  nulls_3                   rounds  nulls_4  
0         company_permalink        0        company_permalink      0.0  
1              company_name        1             company_name      1.0  
2     company_category_list     3632    company_category_list   3410.0  
3      company_country_code    10521     company_country_code   8678.0  
4        company_state_code    13431       company_state_code  10946.0  
5            company_region    12485           company_region  10167.0  
6              company_city    12483             company_city  10164.0  
7        investor_permalink       12  funding_round_permalink      0.0  
8             investor_name       12       funding_round_type      0.0  
9     investor_country_code    30939       funding_round_code  83809.0  
10      investor_state_code    44739                funded_at      0.0  
11          investor_region    43217        raised_amount_usd  19990.0  
12            investor_city    32164                               NaN  
13  funding_round_permalink        0                               NaN  
14       funding_round_type        0                               NaN  
15       funding_round_code    87585                               NaN  
16                funded_at        0                               NaN  
17        raised_amount_usd    21619                               NaN

### `dropna()` for columns with few missing values

Notice that there is one row that doesn't contain a company name, <br>
so let's validate that it is the same company across all datasets, <br>
so we can remove it:

In [150]:
datasetsCols(dfAll, lstCsvNames)

acquisitions              companies              investments  \
0        company_permalink      company_permalink        company_permalink   
1             company_name           company_name             company_name   
2    company_category_list           homepage_url    company_category_list   
3     company_country_code  company_category_list     company_country_code   
4       company_state_code      funding_total_usd       company_state_code   
5           company_region                 status           company_region   
6             company_city   company_country_code             company_city   
7       acquirer_permalink     company_state_code       investor_permalink   
8            acquirer_name         company_region            investor_name   
9   acquirer_category_list           company_city    investor_country_code   
10   acquirer_country_code         funding_rounds      investor_state_code   
11     acquirer_state_code             founded_at          investor_region   
12         acquirer_region       first_funding_at            investor_city   
13           acquirer_city        last_funding_at  funding_round_permalink   
14             acquired_at                              funding_round_type   
15          acquired_month                              funding_round_code   
16            price_amount                                       funded_at   
17     price_currency_code                               raised_amount_usd   

                     rounds  
0         company_permalink  
1              company_name  
2     company_category_list  
3      company_country_code  
4        company_state_code  
5            company_region  
6              company_city  
7   funding_round_permalink  
8        funding_round_type  
9        funding_round_code  
10                funded_at  
11        raised_amount_usd  
12                           
13                           
14                           
15                           
16                           
17

In [151]:
#company_permalink of each row that doesn't have a company name:
for df in dfAll:
    print(df[df['company_name'].isnull()]['company_permalink'].values)

['mendatio"']
['/organization/tell-it-in']
['/organization/tell-it-in']
['/organization/tell-it-in']


So apparently, the `acquisitions` dataset has a different permalink that has no company name, <br>
So we could remove these two companies, as the total number of companies are big enough to get the insight that we want.

`dfAllNulls` dataframe also shows that there are missing values for `acquirer_name` and `investor_name` <br>
which can't be imputed, so we'll also remove them.

Finally, we'll remove the missing values for columns that have less than 100 `NaN`, <br>
as they're small in comparison to their respective datasets:

In [152]:
for df in dfAll:
    nulls = df.isnull().sum()
    cols = nulls[(nulls <= 100) & (nulls != 0)].index.tolist()
    df.dropna(subset=cols, inplace=True)
dfAllNulls = datasetsNulls(dfAll, lstCsvNames)
dfAllNulls

acquisitions  nulls_1              companies  nulls_2  \
0        company_permalink        0      company_permalink      0.0   
1             company_name        0           company_name      0.0   
2    company_category_list     4977           homepage_url   5058.0   
3     company_country_code     4661  company_category_list   3147.0   
4       company_state_code     4940      funding_total_usd      0.0   
5           company_region     4902                 status      0.0   
6             company_city     4902   company_country_code   6956.0   
7       acquirer_permalink        0     company_state_code   8545.0   
8            acquirer_name        0         company_region   8028.0   
9   acquirer_category_list     2387           company_city   8026.0   
10   acquirer_country_code     1168         funding_rounds      0.0   
11     acquirer_state_code     1432             founded_at  15218.0   
12         acquirer_region     1356       first_funding_at      0.0   
13           acquirer_city     1356        last_funding_at      0.0   
14             acquired_at        0                             NaN   
15          acquired_month        0                             NaN   
16            price_amount    13941                             NaN   
17     price_currency_code        0                             NaN   

                investments  nulls_3                   rounds  nulls_4  
0         company_permalink        0        company_permalink      0.0  
1              company_name        0             company_name      0.0  
2     company_category_list     3632    company_category_list   3410.0  
3      company_country_code    10521     company_country_code   8678.0  
4        company_state_code    13431       company_state_code  10946.0  
5            company_region    12485           company_region  10167.0  
6              company_city    12483             company_city  10164.0  
7        investor_permalink        0  funding_round_permalink      0.0  
8             investor_name        0       funding_round_type      0.0  
9     investor_country_code    30927       funding_round_code  83808.0  
10      investor_state_code    44727                funded_at      0.0  
11          investor_region    43205        raised_amount_usd  19990.0  
12            investor_city    32152                               NaN  
13  funding_round_permalink        0                               NaN  
14       funding_round_type        0                               NaN  
15       funding_round_code    87576                               NaN  
16                funded_at        0                               NaN  
17        raised_amount_usd    21619                               NaN

## Converting & Imputing Dates

[SAY WHAT YOU WILL DO HERE FARAH:]

In [153]:
#Farah: plz impute (not drop) the columns' missing values related to dates in each dataset which are:
#founded_at.. and that's it :] 
# Suggestion: convert the dates of all the columns that have date values from "object" to datetime
#       Helpful link: https://pandas.pydata.org/pandas-docs/stable/user_guide/timeseries.html#converting-to-timestamps
# Then, make the founded_at a couple of days before first_funding_at

#Sincerely No. -Farah :)

In [154]:
#farah stop point here

## Cleaning & Imputing Geographical Locations

### Removing `company_state_code` and `company_region`

We don't need `company_state_code` in our analysis, <br>
As country code and city are sufficient to know <br>
an approximation of the geographical location of the company. <br>
Therefore, let's remove it:

In [155]:
companies[['company_country_code', 'company_state_code', 'company_city']].head(10)

company_country_code company_state_code   company_city
0                  IND                 16         Mumbai
1                  USA                 DE  Delaware City
2                  NaN                NaN            NaN
3                  CHN                 22        Beijing
4                  USA                 IL      Champaign
5                  HKG                NaN      Hong Kong
6                  CAN                 BC      Vancouver
7                  USA                 CA  Mountain View
8                  USA                 CA  San Francisco
9                  CHL                 12     Las Condes

In [156]:
companies.drop('company_state_code', axis=1, inplace=True)

`company_region` is also redundent, as it is usually the same as `company_city`.<br>
Let's validate this:

In [157]:
companies[companies.company_region == companies.company_city][['company_region', 'company_city']]

company_region company_city
0             Mumbai       Mumbai
3            Beijing      Beijing
5          Hong Kong    Hong Kong
6          Vancouver    Vancouver
11           Orlando      Orlando
...              ...          ...
66354           Bath         Bath
66360      Milwaukee    Milwaukee
66362         London       London
66363        Beijing      Beijing
66364          Split        Split

[25360 rows x 2 columns]

Let's see the mean of similarity between the strings using builtin library `difflib.SequenceMatcher`:

In [158]:
from difflib import SequenceMatcher
similarities = []
rgs = companies.company_region.values.tolist()
cts = companies.company_city.values.tolist()
for i in range(min(len(rgs), len(cts))):
    if not (rgs[i] != rgs[i] or cts[i] != cts[i]): # If either of the elements is NaN, then don't find the similarity
        similarities.append(SequenceMatcher(None, rgs[i], cts[i]).ratio()) # Gets the similarity between each two strings
avgSim = sum(similarities) / len(similarities)
avgSim

0.6175093624179544

However, `SequenceMatcher` only finds the longest contiguous matching subsequence, <br>
and all other characters will drastically decrease the similarity, which isn't very accurate, <br>
Therefore, let's try to see how many times a string is a substring of another:

In [159]:
subsCount = 0
for i in range(min(len(rgs), len(cts))):
    if not (rgs[i] != rgs[i] or cts[i] != cts[i]):
        subsCount += (rgs[i] in cts[i] or cts[i] in rgs[i])
subsCount

29884

So to summarize: almost half of the companies have the same data for `region` and `city` (25360) <br>
There are around 5000 companies that have the same semantic meaning between the data <br>
(e.g: 'New York City' is the same as 'NYC, New York City') <br>
Therefore, `company_region` can be removed without losing any possible future insights:

In [160]:
companies.drop('company_region', axis=1, inplace=True)

### Imputing `company_country_code` and `company_city`

A way to impute country codes is to do the following:
1. check the company's permalink
2. find HTML that contains 

In [161]:
companies[companies.company_country_code.isnull() == False].sample(10)

company_permalink               company_name  \
44935                  /organization/premitech                  PremiTech   
5377                   /organization/aw-energy                  AW-Energy   
39138                    /organization/nexgrid                    NEXGRID   
33200                       /organization/loom                       Loom   
45314                    /organization/prodeaf                    ProDeaf   
672    /organization/abaad-embodied-design-llc  Abaad Embodied Design LLC   
5080       /organization/autifony-therapeutics      Autifony Therapeutics   
4754                    /organization/athersys                   Athersys   
51180                   /organization/shipbeat                   shipbeat   
14960                   /organization/dealbird                   DealBird   

                   homepage_url  \
44935  http://www.premitech.com   
5377       http://aw-energy.com   
39138       http://smartgrid.us   
33200       http://www.loom.com   
45314    http://www.prodeaf.net   
672         http://www.abaad.ae   
5080    http://www.autifony.com   
4754    http://www.athersys.com   
51180       http://shipbeat.com   
14960   http://www.dealbird.com   

                                   company_category_list funding_total_usd  \
44935                                           Software           4010000   
5377    Clean Energy|Clean Technology|Renewable Energies          19427954   
39138                                Enterprise Software           2250046   
33200  Cloud Computing|Flash Storage|Mobile|Photograp...           1400000   
45314  Communications Infrastructure|Service Provider...            142627   
672                                 3D|Rapidly Expanding                 -   
5080                                       Biotechnology          29769365   
4754                                       Biotechnology           9000002   
51180        Delivery|Developer APIs|E-Commerce|Shipping           1592388   
14960  Automotive|Coupons|Discounts|Group Buying|Reta...                 -   

          status company_country_code   company_city  funding_rounds  \
44935     closed                  DNK     Hedehusene               1   
5377   operating                  FIN         Vantaa               2   
39138  operating                  USA   Locust Grove               6   
33200   acquired                  USA  San Francisco               1   
45314  operating                  BRA         Recife               2   
672    operating                  ARE      Abu Dhabi               1   
5080   operating                  GBR         London               4   
4754         ipo                  USA      Cleveland               1   
51180  operating                  DNK     Copenhagen               1   
14960     closed                  USA   Santa Monica               1   

      founded_at first_funding_at last_funding_at  
44935   1/1/1999        1/29/2007       1/29/2007  
5377    1/1/2002         1/1/2004       9/16/2014  
39138   1/1/2008        7/26/2010      11/13/2013  
33200        NaN        10/4/2013       10/4/2013  
45314   1/1/2010         5/1/2013        8/1/2013  
672     1/1/2012         1/1/2012        1/1/2012  
5080    1/1/2011        8/22/2011       7/10/2015  
4754    1/1/1995        3/23/2012       3/23/2012  
51180   1/1/2014        8/22/2014       8/22/2014  
14960  6/10/2010         1/1/2010        1/1/2010

In [162]:
companies[companies.company_country_code.isnull() == False].sample(5)

company_permalink      company_name  \
3694              /organization/appia             Appia   
9652     /organization/carbon-objects    Carbon Objects   
18725  /organization/erms-corporation  ERMS Corporation   
52650               /organization/smx               SMX   
39356       /organization/nightingale       Nightingale   

                        homepage_url  \
3694            http://www.appia.com   
9652                             NaN   
18725            http://ermscorp.com   
52650        http://www.smxemail.com   
39356  http://www.nightingaleapp.com   

                                   company_category_list funding_total_usd  \
3694                       Advertising|App Stores|Mobile          44250000   
9652              Design|Entertainment|Graphics|Printing            250000   
18725                                           Software           3500000   
52650                                           Security           1670000   
39356  Analytics|Health and Wellness|Health Care Info...            780000   

          status company_country_code   company_city  funding_rounds  \
3694    acquired                  USA         Durham               8   
9652   operating                  USA    Chattanooga               1   
18725  operating                  CAN        Toronto               1   
52650  operating                  NZL       Auckland               1   
39356  operating                  USA  San Francisco               3   

      founded_at first_funding_at last_funding_at  
3694    1/1/2008         6/1/2008      11/14/2013  
9652         NaN        1/10/2014       1/10/2014  
18725   1/1/2002        9/12/2013       9/12/2013  
52650   2/1/2006        7/20/2009       7/20/2009  
39356  7/23/2013        7/23/2013       3/24/2015

In [163]:
companies.company_country_code.unique()

array(['IND', 'USA', nan, 'CHN', 'HKG', 'CAN', 'CHL', 'GBR', 'FRA', 'AUS',
       'DNK', 'ROM', 'AUT', 'KOR', 'NLD', 'SWE', 'JPN', 'RUS', 'SGP',
       'NOR', 'COL', 'ESP', 'BEL', 'IRL', 'ITA', 'ISR', 'NZL', 'CZE',
       'DEU', 'CHE', 'BRA', 'BGR', 'SVN', 'JOR', 'HUN', 'BWA', 'NGA',
       'FIN', 'TUR', 'IDN', 'ARE', 'CYP', 'POL', 'CRI', 'PRT', 'ARG',
       'TWN', 'KHM', 'THA', 'SVK', 'UKR', 'LTU', 'ISL', 'ZAF', 'MEX',
       'VEN', 'URY', 'MUS', 'KEN', 'PHL', 'MNE', 'VNM', 'GHA', 'PSE',
       'MYS', 'PER', 'EGY', 'PAN', 'LVA', 'GGY', 'ALB', 'UGA', 'HRV',
       'EST', 'LBN', 'GRC', 'PAK', 'NPL', 'LUX', 'SAU', 'IRN', 'GTM',
       'BAH', 'ARM', 'BGD', 'MDA', 'SRB', 'TUN', 'ZWE', 'BRB', 'NIC',
       'TAN', 'PRI', 'TTO', 'BHR', 'CMR', 'BLR', 'CYM', 'BRN', 'ECU',
       'SLV', 'MLT', 'ZMB', 'MKD', 'GIB', 'TGO', 'LAO', 'BMU', 'HND',
       'MCO', 'MMR', 'KAZ', 'QAT', 'DOM', 'MAR', 'LIE', 'GEO', 'MOZ',
       'DZA', 'JAM', 'KNA', 'AZE', 'SEN', 'RWA', 'DMA', 'UZB', 'BLZ',
       'OMN', '

## Checking All `dtypes`

In [164]:
def datasetsDtypes(dfList, datasetNames, dtypesCol="dtypes_", pad=""):
    cols = []
    nulls = []
    maxArrayLen = 0
    for i, df in enumerate(dfList):
        cols.append(df.columns.tolist())
        nulls.append(df.dtypes.tolist())
        maxArrayLen = max(maxArrayLen, len(cols[i]))
    
    dictCsvs = {}
    for i, df in enumerate(dfList):
        cols[i] += [pad] * (maxArrayLen - len(cols[i])) #padding the lists to make them have equal lengths 
        nulls[i] += [None] *  (maxArrayLen - len(nulls[i]))
        dictCsvs[datasetNames[i]] = cols[i]
        dictCsvs[dtypesCol + str(i+1)] = nulls[i]
    
    return pd.DataFrame(dictCsvs)

In [165]:
datasetsDtypes(dfAll, lstCsvNames)

acquisitions dtypes_1              companies dtypes_2  \
0        company_permalink   object      company_permalink   object   
1             company_name   object           company_name   object   
2    company_category_list   object           homepage_url   object   
3     company_country_code   object  company_category_list   object   
4       company_state_code   object      funding_total_usd   object   
5           company_region   object                 status   object   
6             company_city   object   company_country_code   object   
7       acquirer_permalink   object           company_city   object   
8            acquirer_name   object         funding_rounds    int64   
9   acquirer_category_list   object             founded_at   object   
10   acquirer_country_code   object       first_funding_at   object   
11     acquirer_state_code   object        last_funding_at   object   
12         acquirer_region   object                            None   
13           acquirer_city   object                            None   
14             acquired_at   object                            None   
15          acquired_month   object                            None   
16            price_amount  float64                            None   
17     price_currency_code   object                            None   

                investments dtypes_3                   rounds dtypes_4  
0         company_permalink   object        company_permalink   object  
1              company_name   object             company_name   object  
2     company_category_list   object    company_category_list   object  
3      company_country_code   object     company_country_code   object  
4        company_state_code   object       company_state_code   object  
5            company_region   object           company_region   object  
6              company_city   object             company_city   object  
7        investor_permalink   object  funding_round_permalink   object  
8             investor_name   object       funding_round_type   object  
9     investor_country_code   object       funding_round_code   object  
10      investor_state_code   object                funded_at   object  
11          investor_region   object        raised_amount_usd  float64  
12            investor_city   object                              None  
13  funding_round_permalink   object                              None  
14       funding_round_type   object                              None  
15       funding_round_code   object                              None  
16                funded_at   object                              None  
17        raised_amount_usd  float64                              None

In [166]:
#AISHA: see what you can do with the dtypes above (except dates, that's on Farah) ^^

In [167]:
print("Aisha is happy ^^")

Aisha is happy ^^


In [168]:
#Aisha stop

## `rounds` Dataset

In [169]:
dfAllCols.iloc[:, -2:]

investments                   rounds
0         company_permalink        company_permalink
1              company_name             company_name
2     company_category_list    company_category_list
3      company_country_code     company_country_code
4        company_state_code       company_state_code
5            company_region           company_region
6              company_city             company_city
7        investor_permalink  funding_round_permalink
8             investor_name       funding_round_type
9     investor_country_code       funding_round_code
10      investor_state_code                funded_at
11          investor_region        raised_amount_usd
12            investor_city                         
13  funding_round_permalink                         
14       funding_round_type                         
15       funding_round_code                         
16                funded_at                         
17        raised_amount_usd

In [170]:
invests.shape, rounds.shape

((168634, 18), (114948, 12))

In [171]:
invests.head()

company_permalink        company_name company_category_list  \
0             /organization/0-6-com             0-6.com           Curated Web   
1    /organization/004-technologies    004 Technologies              Software   
2  /organization/01games-technology  01Games Technology                 Games   
3              /organization/0xdata              H2O.ai             Analytics   
4              /organization/0xdata              H2O.ai             Analytics   

  company_country_code company_state_code         company_region  \
0                  CHN                 22                Beijing   
1                  USA                 IL  Springfield, Illinois   
2                  HKG                NaN              Hong Kong   
3                  USA                 CA            SF Bay Area   
4                  USA                 CA            SF Bay Area   

    company_city                    investor_permalink  \
0        Beijing     /organization/dt-capital-partners   
1      Champaign       /organization/venturecapital-de   
2      Hong Kong     /organization/cyberport-hong-kong   
3  Mountain View             /organization/capital-one   
4  Mountain View  /organization/nexus-venture-partners   

            investor_name investor_country_code investor_state_code  \
0     DT Capital Partners                   CHN                  23   
1   VCDE Venture Partners                   DEU                   5   
2     Cyberport Hong Kong                   HKG                 NaN   
3             Capital One                   USA                  VA   
4  Nexus Venture Partners                   USA                  CA   

    investor_region investor_city  \
0          Shanghai      Shanghai   
1         Frankfurt     Frankfurt   
2         Hong Kong     Hong Kong   
3  Washington, D.C.        Mclean   
4       SF Bay Area    Menlo Park   

                           funding_round_permalink funding_round_type  \
0  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   
1  /funding-round/1278dd4e6a37fa4b7d7e06c21b3c1830            venture   
2  /funding-round/7d53696f2b4f607a2f2a8cbb83d01839        undisclosed   
3  /funding-round/3bb2ee4a2d89251a10aaa735b1180e44            venture   
4  /funding-round/383a9bd2c04f7038bb543ccef5ba3eae               seed   

  funding_round_code  funded_at  raised_amount_usd  
0                  A  3/19/2008          2000000.0  
1                NaN  7/24/2014                NaN  
2                NaN   7/1/2014            41250.0  
3                  B  11/9/2015         20000000.0  
4                NaN  5/22/2013          3000000.0

In [172]:
rounds.head()

company_permalink            company_name  \
0                  /organization/-fame                   #fame   
1               /organization/-qounter                :Qounter   
2               /organization/-qounter                :Qounter   
3  /organization/-the-one-of-them-inc-  (THE) ONE of THEM,Inc.   
4                /organization/0-6-com                 0-6.com   

                               company_category_list company_country_code  \
0                                              Media                  IND   
1  Application Platforms|Real Time|Social Network...                  USA   
2  Application Platforms|Real Time|Social Network...                  USA   
3                                  Apps|Games|Mobile                  NaN   
4                                        Curated Web                  CHN   

  company_state_code company_region   company_city  \
0                 16         Mumbai         Mumbai   
1                 DE     DE - Other  Delaware City   
2                 DE     DE - Other  Delaware City   
3                NaN            NaN            NaN   
4                 22        Beijing        Beijing   

                           funding_round_permalink funding_round_type  \
0  /funding-round/9a01d05418af9f794eebff7ace91f638            venture   
1  /funding-round/22dacff496eb7acb2b901dec1dfe5633            venture   
2  /funding-round/b44fbb94153f6cdef13083530bb48030               seed   
3  /funding-round/650b8f704416801069bb178a1418776b            venture   
4  /funding-round/5727accaeaa57461bd22a9bdd945382d            venture   

  funding_round_code   funded_at  raised_amount_usd  
0                  B    1/5/2015         10000000.0  
1                  A  10/14/2014                NaN  
2                NaN    3/1/2014           700000.0  
3                  B   1/30/2014          3406878.0  
4                  A   3/19/2008          2000000.0

In [173]:
len(invests["company_name"].unique()), len(rounds["company_name"].unique())

(44577, 66102)

Since there are companies in `rounds` dataset that are not in `investments` dataset, <br>
therefore we should keep `rounds` dataset for further analysis.

## `companies` Dataset

In [174]:
companies.sample(5).T

43270                          64271  \
company_permalink      /organization/pheedo  /organization/womensforum-com   
company_name                         Pheedo                    Womensforum   
homepage_url                            NaN    http://www.womensforum.com/   
company_category_list           Advertising                         Forums   
funding_total_usd                   6600000                        7000000   
status                            operating                      operating   
company_country_code                    USA                            USA   
company_city                        Oakland                        Chicago   
funding_rounds                            2                              1   
founded_at                              NaN                       1/1/1996   
first_funding_at                   3/1/2005                      3/31/2000   
last_funding_at                    6/1/2006                      3/31/2000   

                                       65986                            53283  \
company_permalink       /organization/ziibra            /organization/somatix   
company_name                          ZIIBRA                          Somatix   
homepage_url           http://www.ziibra.com        http://www.somatixinc.com   
company_category_list         CRM|E-Commerce  Analytics|Health Care|Wearables   
funding_total_usd                     318000                           750000   
status                             operating                           closed   
company_country_code                     USA                              ISR   
company_city                         Seattle                         Herzliya   
funding_rounds                             3                                1   
founded_at                          1/1/2012                         2/1/2015   
first_funding_at                    5/8/2013                        4/22/2015   
last_funding_at                    7/27/2015                        4/22/2015   

                                           37789  
company_permalink       /organization/myneighbor  
company_name                          MyNeighbor  
homepage_url           http://www.MyNeighbor.com  
company_category_list               Marketplaces  
funding_total_usd                         354573  
status                                 operating  
company_country_code                         USA  
company_city                             Seattle  
funding_rounds                                 1  
founded_at                              4/1/2014  
first_funding_at                       9/10/2015  
last_funding_at                        9/10/2015

In [175]:
companies.isnull().sum()

company_permalink            0
company_name                 0
homepage_url              5058
company_category_list     3147
funding_total_usd            0
status                       0
company_country_code      6956
company_city              8026
funding_rounds               0
founded_at               15218
first_funding_at             0
last_funding_at              0
dtype: int64

we will need to remove the none values of the following columns <br>
based on the question that we are trying to gain insight for: <br>
1.	Can the factors that affect a start-up’s growth be determined?
    -	For this: We are initially interested in <br> `funding_total_usd`, `status`, `funding_rounds`, `first_funding_at`, and `last_funding_at`
    -   So we could remove the 24 null values of `first_funding_at` since they are so small <br>
        compared to the rest of the dataset
2.  Do taxes play an important factor in deciding a start-up’s success?

